File: framework/languages/node/e2e/ToActorMessaging/Shared/messages.ts
import { ZLinkPacket } from '@zlink-systems/framework';

export const PacketNames = {
  actorMsg: 'ActorMsg',
  actorReq: 'ActorReq',
  actorPushReq: 'ActorPushReq',
  actorPushNotify: 'ActorPushNotify',
  bindActorReq: 'BindActorReq'
} as const;

export class ActorMsg {
  constructor(
    readonly scenario: string,
    readonly actorId: string,
    readonly value: string
  ) {}
}

export class ActorReq {
  constructor(
    readonly scenario: string,
    readonly actorId: string,
    readonly value: string
  ) {}
}

export interface ActorRes {
  readonly scenario: string;
  readonly actorId: string;
  readonly value: string;
}

export class ActorPushReq {
  constructor(
    readonly scenario: string,
    readonly actorId: string,
    readonly value: string
  ) {}
}

@ZLinkPacket(PacketNames.actorPushNotify)
export class ActorPushNotify {
  constructor(
    readonly scenario: string,
    readonly actorId: string,
    readonly value: string
  ) {}
}

export interface ActorEvidence {
  readonly scenario: string;
  readonly actorId: string;
  readonly kind: string;
  readonly value: string;
}

export interface ActorRefPayload {
  readonly nodeRid: string;
  readonly actorId: string;
  readonly objectGeneration: string;
  readonly meshName: string;
}

export interface ActorEnsureRes {
  readonly actorId: string;
  readonly actor: ActorRefPayload;
}

export interface BindActorReq {
  readonly actor: ActorRefPayload;
}

export interface BindActorRes {
  readonly actorId: string;
  readonly nodeRid: string;
  readonly objectGeneration: string;
  readonly boundCount: number;
}

export interface SessionBindingRes {
  readonly actorId: string;
  readonly sessionIds: readonly string[];
}

export interface ActorCallReq {
  readonly scenario: string;
  readonly actorId: string;
  readonly actor?: ActorRefPayload;
  readonly value: string;
}

export interface ActorCallRes {
  readonly scenario: string;
  readonly actorId: string;
  readonly result: string;
  readonly errorKind?: string;
}

export function actorMsg(scenario: string, actorId: string, value: string): ActorMsg {
  return new ActorMsg(scenario, actorId, value);
}

export function actorReq(scenario: string, actorId: string, value: string): ActorReq {
  return new ActorReq(scenario, actorId, value);
}

export function actorPush(scenario: string, actorId: string, value: string): ActorPushReq {
  return new ActorPushReq(scenario, actorId, value);
}

File: framework/languages/dotnet/e2e/ToActorMessaging/Shared/Messages.cs
namespace ToActorMessaging.Shared;

public sealed record ActorMsg(string Scenario, string ActorId, string Value);

public sealed record ActorReq(string Scenario, string ActorId, string Value);

public sealed record ActorRes(string Scenario, string ActorId, string Value);

public sealed record ActorEvidence(
    string Scenario,
    string ActorId,
    string Kind,
    string Value,
    string? NodeRid = null,
    ulong? Generation = null,
    string? PacketName = null,
    string? RequestId = null);

public sealed record ActorCallReq(
    string Scenario,
    string ActorId,
    string Value,
    string? TargetNodeRid = null,
    ulong? TargetGeneration = null);

public sealed record ActorCallRes(string Scenario, string ActorId, string Result, string? ErrorKind = null);

public sealed record DestroyActorReq(string ActorId, string Scenario);

public sealed record DestroyActorRes(string ActorId, bool Destroyed);

public sealed record BindActorReq(string ActorId);

public sealed record BindActorRes(string ActorId, string NodeRid, ulong Generation);

public sealed record BoundPushReq(string Scenario, string ActorId, string Value);

public sealed record BoundPushRes(
    string ActorId,
    string Value,
    bool Submitted,
    string? ErrorKind = null);

public sealed record BoundPushNotify(string Scenario, string ActorId, string Value);

public sealed record ActorRouteStatus(string ActorId, bool Exists);

public sealed record BoundSessionSnapshot(
    string ActorId,
    string? SessionRid,
    long Revision);
